File: src/main/java/seremis/geninfusion/soul/InstanceHelper.java
package seremis.geninfusion.soul;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import seremis.geninfusion.api.soul.*;
import seremis.geninfusion.api.soul.lib.Genes;
import seremis.geninfusion.soul.allele.AlleleBoolean;
import seremis.geninfusion.soul.entity.EntitySoulCustom;
import seremis.geninfusion.soul.entity.EntitySoulCustomCreature;

import java.lang.reflect.Constructor;

/**
 * @author Seremis
 */
public class InstanceHelper implements IInstanceHelper {

    @Override
    public IEntitySoulCustom getSoulEntityInstance(World world, ISoul soul, double x, double y, double z) {
        IEntitySoulCustom entity = new EntitySoulCustom(world, soul, x, y, z);

        if(((AlleleBoolean)SoulHelper.geneRegistry.getActiveFor(entity, Genes.GENE_IS_CREATURE)).value) {
            entity = new EntitySoulCustomCreature(world, soul, x, y, z);
        }
        return entity;
    }

    @Override
    public ISoul getISoulInstance(IChromosome[] chromosomes) {
        return new Soul(chromosomes);
    }

    @Override
    public IChromosome getIChromosomeInstance(IAllele allele1, IAllele allele2) {
        return new Chromosome(allele1, allele2);
    }

    @Override
    public IAllele getIAlleleInstance(EnumAlleleType alleleType, Object... args) {
        try {
            Constructor<?> ctor = alleleType.clazz.getConstructor(NBTTagCompound.class);
            Object object = ctor.newInstance(new Object[] { args });

            return (IAllele) object;
        } catch(Exception e) {
            e.printStackTrace();
        }

        return null;
    }
}

File: src/main/scala/seremis/geninfusion/soul/InstanceHelper.scala
package seremis.geninfusion.soul

import net.minecraft.entity.{EntityLiving, EntityList}
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import seremis.geninfusion.api.soul._
import seremis.geninfusion.api.soul.lib.Genes
import seremis.geninfusion.api.util.AncestryNode
import seremis.geninfusion.soul.entity.{EntitySoulCustom, EntitySoulCustomAgeable, EntitySoulCustomCreature}

class InstanceHelper extends IInstanceHelper {

    override def getSoulEntityInstance(world: World, soul: ISoul, x: Double, y: Double, z: Double): IEntitySoulCustom = {
        var entity: IEntitySoulCustom = null

        if(SoulHelper.geneRegistry.getValueFromAllele(entity, Genes.GeneIsCreature)) {
            entity = new EntitySoulCustomCreature(world, soul, x, y, z)
        }

        if(SoulHelper.geneRegistry.getValueFromAllele(entity, Genes.GeneCanProcreate)) {
            entity = new EntitySoulCustomAgeable(world, soul, x, y, z)
        }

        if(entity == null) {
            entity = new EntitySoulCustom(world, soul, x, y, z)
        }

        entity
    }

    override def getISoulInstance(chromosomes: Array[IChromosome], name: Option[String], ancestry: AncestryNode): Option[ISoul] = Some(new Soul(chromosomes, name, ancestry))

    override def getIChromosomeInstance(geneName: String, allele1: IAllele, allele2: IAllele): IChromosome = new Chromosome(geneName, allele1, allele2)

    override def getIAlleleInstance(args: AnyRef*): IAllele =  {
        //TODO test this
        new Allele(args(0).asInstanceOf[Boolean], args(1), SoulHelper.alleleTypeRegistry.getAlleleTypeForClass(args(1).getClass))
    }

    override def getSoulEntityInstance(compound: NBTTagCompound, world: World): IEntitySoulCustom = {
        val entity = getSoulEntityInstance(world, new Soul(compound), 0, 0, 0)
        entity.asInstanceOf[EntityLiving].readFromNBT(compound)

        entity
    }

    override def getIAlleleInstance(compound: NBTTagCompound): IAllele = {
        new Allele(compound)
    }

    override def getISoulInstance(compound: NBTTagCompound): Option[ISoul] = {
        Some(new Soul(compound))
    }

    override def getIChromosomeInstance(compound: NBTTagCompound): IChromosome = {
        new Chromosome(compound)
    }
}
